src/system.rs

```diff
@@ -1197,6 +1197,162 @@ macro_rules! system {
+        pub mod fmt {
+            use $crate::lib::fmt;
+            use super::{Dimension, Quantity, Unit, Units, from_base};
+            use $crate::num::Num;
+            use $crate::Conversion;
+
+            /// Specifies the display style to use.
+            #[derive(Clone, Copy, Debug)]
+            #[allow(dead_code)] // We don't use this internally.
+            pub enum DisplayStyle {
+                /// Display the value and a unit abbreviation, e.g. "1.0 m", "327 s".
+                Abbreviation,
+                /// Display the value and full unit name (pluralized as appropriate),
+                /// e.g. "1 kilogram", "756 feet".
+                Description,
+            }
+
+            /// Provides an interface to specify a display style and unit.
+            ///
+            /// # Usage
+            /// ## Indirect style
+            /// ```
+            /// # use uom::si::Unit;
+            /// # use uom::si::f32::*;
+            /// # use uom::si::length::{centimeter, meter};
+            /// # use uom::si::fmt::Arguments;
+            /// # use uom::si::fmt::Indirect;
+            /// # use uom::si::fmt::DisplayStyle::*;
+            /// let length = Length::new::<meter>(1.0);
+            /// let a = Arguments::new(centimeter, Description);
+            /// assert_eq!("100 centimeters", format!("{}", a.with(length)));
+            /// ```
+            /// ## Direct style
+            /// ```
+            /// # use uom::si::Unit;
+            /// # use uom::si::f32::*;
+            /// # use uom::si::length::{centimeter, meter};
+            /// # use uom::si::fmt::Arguments;
+            /// # use uom::si::fmt::Direct;
+            /// # use uom::si::fmt::DisplayStyle::*;
+            /// let length = Length::new::<meter>(1.0);
+            /// let a = Arguments::new(length, centimeter, Description);
+            /// assert_eq!("100 centimeters", format!("{}", a));
+            /// ```
+            #[allow(missing_debug_implementations)] // Prevent accidental direct use.
+            #[derive(Clone, Copy)]
+            pub struct Arguments<T: Unit> {
+                subunit: T,
+                style: DisplayStyle,
+            }
+
+            /// Allows for indirect, delayed formatting of an arbitrary number of quantities
+            /// in the specified style and unit.
+            pub trait Indirect<T>
+            where
+                T: Unit {
+                /// Creates an instance of `Arguments<T>` that can be used to format quantites
+                /// later.
+                ///
+                /// To format a quantity, use the `with` method on `Arguments<T>` to obtain
+                /// a displayable object.
+                fn new(subunit: T, style: DisplayStyle) -> Arguments<T>;
+            }
+
+            /// Allows for direct and immediate formatting of quantities in the specified style
+            /// and unit.
+            pub trait Direct<T>
```

Comment:
I'm thinking that using traits to overload `new()` adds too much complexity and just having two methods: `new(...) -> Arguments` and `new_with(...) -> QuantityArguments` makes things easier to use. With traits the trait needs to be in scope  and if you happen to have both in scope you need to use fully qualified method call syntax.